program/rust/src/rust_oracle.rs

```diff
@@ -44,27 +56,138 @@ use crate::utils::pyth_assert;
 
 use super::c_entrypoint_wrapper;
 
+const PRICE_T_SIZE: usize = size_of::<pc_price_t>();
+const PRICE_ACCOUNT_SIZE: usize = size_of::<PriceAccountWrapper>();
+
+
 ///Calls the c oracle update_price, and updates the Time Machine if needed
 pub fn update_price(
     _program_id: &Pubkey,
-    _accounts: &[AccountInfo],
+    accounts: &[AccountInfo],
     _instruction_data: &[u8],
     input: *mut u8,
 ) -> OracleResult {
-    //For now, we did not change the behavior of this. this is just to show the proposed structure
-    // of the program
-    c_entrypoint_wrapper(input)
+    let price_account_info = &accounts[1];
+    let account_len = price_account_info.try_data_len()?;
+    if account_len < PRICE_ACCOUNT_SIZE {
+        if account_len != PRICE_T_SIZE {
+            return Err(ProgramError::InvalidArgument);
+        }
+        msg!("Please resize the account to allow for SMA tracking!");
+        return c_entrypoint_wrapper(input);
+    }
+    let c_ret_value = c_entrypoint_wrapper(input)?;
+    if c_ret_value == SUCCESSFULLY_UPDATED_AGGREGATE {
+        let mut price_account = load_account_as_mut::<PriceAccountWrapper>(price_account_info)?;
+        price_account.add_price()?;
+        msg!("updated tracker!");
+    }
+    Ok(c_ret_value)
+}
+
+///resize price account and initialize the TimeMachineStructure
+#[cfg(feature = "resize-account")]
+pub fn upgrade_price_account<'a>(
+    funding_account_info: &AccountInfo<'a>,
+    price_account_info: &AccountInfo<'a>,
+    system_program: &AccountInfo<'a>,
+    program_id: &Pubkey,
+) -> OracleResult {
+    pyth_assert(
+        valid_funding_account(&funding_account_info),
+        OracleError::InvalidFundingAccount.into(),
+    )?;
+
+    pyth_assert(
+        valid_writable_account(program_id, &price_account_info, size_of::<pc_price_t>()),
+        OracleError::InvalidWritableAccount.into(),
+    )?;
+
+    let account_len = price_account_info.try_data_len()?;
+    //const PRICE_T_SIZE: usize = size_of::<pc_price_t>();
+    match account_len {
+        PRICE_T_SIZE => {
+            //compute the number of lamports needed in the price account to update it
```

Comment:
This comment is slightly misleading - it should maybe be "ensure account is still rent exempt after resizing".